### PR TITLE
Add ability to specify validator for entire store

### PR DESCRIPTION
**Pull request by SyntheticDave**
This change allows us to specify validations for an entire store, rather than just a single attribute on the store. This is done by calling `validate` in the specific `Avetmiss::Stores::Validator`.

For example, to call `qualification_min_units_achieved?` on the validator, passing it the store, use the line:

```
validate :qualification_min_units_achieved?
```

If this method returns false, `add_result` will be called with the result:

```
{
    base: store_proxy,
    attribute: :base,
    value: '',
    message: '',
    validator_type: :qualification_min_units_achieved?
}
```

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -79,3 +79,6 @@ DEPENDENCIES
   rake (~> 10.1)
   rspec (~> 2.14)
   simplecov (~> 0.7)
+
+BUNDLED WITH
+   1.10.6
```
Comment (schlick) [inline Gemfile.lock]: I believe a Gemfile.lock shouldn't be included in a gem repo.

File: avetmiss_validations.gemspec [hunk 2]
```diff
@@ -22,8 +22,8 @@ Gem::Specification.new do |s|
   s.add_development_dependency 'byebug', '~> 2.7'
   s.add_development_dependency 'rake', '~> 10.1'
   s.add_development_dependency 'coveralls'
-  s.add_dependency 'activesupport'
-  s.add_dependency 'activemodel'
+  s.add_dependency 'activesupport', '~> 3.2'
+  s.add_dependency 'activemodel', '~> 3.2'
```
Comment (schlick) [inline avetmiss_validations.gemspec]: Why 3.2?

File: lib/avetmiss_validations/active_model/validator.rb [hunk 3]
```diff
@@ -1,11 +1,20 @@
...
   self.attributes_validations = []
 
+  class_attribute :store_validations
+  self.store_validations = []
+
+  # Allows the caller to register validations on a single attibute
   def self.validates(attribute, validations)
     options = validations.extract!(:if, :unless)
-    self.attributes_validations << { attribute: attribute, validations: validations, options: options }
+    self.attributes_validations += [{ attribute: attribute, validations: validations, options: options }]
```
Comment (schlick) [inline lib/avetmiss_validations/active_model/validator.rb]: Interesting change: from `<<` to `+=`. Why?

File: Gemfile.lock (hunk 1 above)
Comment (SyntheticDave) [inline Gemfile.lock]: Ok, I'll remove the lock file from version control.

File: avetmiss_validations.gemspec (hunk 2 above)
Comment (SyntheticDave) [inline avetmiss_validations.gemspec]: I initially updated to that to match what was currently in the lock file. 
I'm going to update both to a range `>= 3, < 5`.

File: lib/avetmiss_validations/active_model/validator.rb (hunk 3 above)
Comment (SyntheticDave) [inline lib/avetmiss_validations/active_model/validator.rb]: This is a gotcha with `class_attributes`. Any class attribute that is a collection should not be modified in place, as the class_attributes of any derived class points to the same object. Modifying the collection using a setter ensures different derived classes don't alter each other's attributes.

I discovered this when creating multiple instances of an anonymous class that extends `AvetmissValidations::ActiveModel::Validator` for testing, and found that each instance doubled up on the stored validations when I created the second.

This behaviour is covered in the [rails documention](http://apidock.com/rails/Class/class_attribute).

Comment (schlick) [inline lib/avetmiss_validations/active_model/validator.rb]: Thanks for the detailed response!

**Comment by schlick** [comment]
👍

**Pull request merged**